src/graph.jl

```diff
@@ -129,32 +78,128 @@ function Base.transpose(g::Graph{loops,T}) where {loops,T}
     end
     B_colptr[1] = 1
 
-    return Graph{loops,T}(n, m, B_colptr, B_rowval)
+    return SparsePatternCSC{T}(n, m, B_colptr, B_rowval)
+end
+
+## Adjacency graph
+
+"""
+    AdjacencyGraph{T}
+
+Undirected graph without self-loops representing the nonzeros of a symmetric matrix (typically a Hessian matrix).
+
+The adjacency graph of a symmetrix matric `A ∈ ℝ^{n × n}` is `G(A) = (V, E)` where
+
+- `V = 1:n` is the set of rows or columns `i`/`j`
+- `(i, j) ∈ E` whenever `A[i, j] ≠ 0` and `i ≠ j`
+
+# Constructors
+
+    AdjacencyGraph(A::SparseMatrixCSC)
+
+# Fields
+
+- `S::SparsePatternCSC{T}`
+
+# References
+
+> [_What Color Is Your Jacobian? SparsePatternCSC Coloring for Computing Derivatives_](https://epubs.siam.org/doi/10.1137/S0036144504444711), Gebremedhin et al. (2005)
+"""
+struct AdjacencyGraph{T}
+    S::SparsePatternCSC{T}
+end
+
+AdjacencyGraph(A::SparseMatrixCSC) = AdjacencyGraph(SparsePatternCSC(A))
+
+pattern(g::AdjacencyGraph) = g.S
+nb_vertices(g::AdjacencyGraph) = pattern(g).n
+vertices(g::AdjacencyGraph) = 1:nb_vertices(g)
+
+function neighbors(g::AdjacencyGraph, v::Integer)
+    S = pattern(g)
+    neighbors_with_loops = view(rowvals(S), nzrange(S, v))
+    return Iterators.filter(!=(v), neighbors_with_loops)  # TODO: optimize
+end
+
+function degree(g::AdjacencyGraph, v::Integer)
+    d = 0
+    for u in neighbors(g, v)
+        if u != v
+            d += 1
+        end
+    end
+    return d
+end
+
+function nb_edges(g::AdjacencyGraph)
+    S = pattern(g)
+    ne = 0
+    for j in vertices(g)
+        for k in nzrange(S, j)
+            i = rowvals(S)[k]
+            if i != j
```

Comment:
Maybe just check that `i < j`. You don't need `ne ÷ 2` at the end with this modification.